### ISSUE TITLE
buildkite: Extract problematic messages from the logs and present them in the buildkite UI

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#17635]
### Feature request

Use https://buildkite.com/docs/agent/v3/cli-annotate to post-process the logs from a build in order to extract important log messages out of them.

1. We have two places where the hook can be installed, ci/plugin/cloudtest/pre-exit and ci/plugin/mzcompose/pre-exit
2. Given that buildkite will bundle all the annotations from all the failing steps/jobs of a build and present them together at the summary page of the build, the individual reports must point back to the test (buildkite step id) that they came from.
3. We are interested in the following:

From Materialize:
```
limits-materialized-1     | thread 'timely:work-0' panicked at 'batch fetcher could not fetch batch part: unexpected missing blob: wf0804f00-4664-4b27-bf41-556ef79a892d/pb08059fa-d472-431f-9a20-0e2d6b10046f for shard: se08d377d-471d-429c-9097-a6d3c200235b', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-02195a46e47a1a796-1/materialize/tests/src/persist-client/src/fetch.rs:205:9
```

From CRDB:

```
internal error:
```

```
* FATAL:
```

From the kernel OOM killer:

```
Feb 13 05:00:19 ip-10-61-109-220.ec2.internal kernel: Out of memory: Killed process 38205 (environmentd) total-vm:14177356kB, anon-rss:9735152kB, file-rss:7044kB, shmem-rss:0kB, UID:999 pgtables:26408kB oom_score_adj:0
```

There are a couple of design decisions to be made:

1. Are we going to fish for specific regexps in specific files? My gut feeling is that we should fish for all regexps in all files because we may decide to move things around and we would want this functionality to not silently break. For example, services.log is already too large for comfort and we may need to split it up going forward. Also, if we ever move our Buildkite agents from Amazon Linux (the Buildkite default) to Ubuntu (the Mz corporate default), then the OOM messages may start appearing in the dmesg output rather than the journalctl output.

2. Where to put the code and what language to write it in. The current two buildkite plugins are bash-only, but actually some basic python is also available on the Buildkite machines (and , a full , up-to-date python is available via the `bin/ci-builder run stable` . Ideally the two plugins should be able to share the detection code. An alternative option would be to put it in the cloudtest and mzcompose frameworks themselves -- currently I believe the junitxml files are being generated within mzcompose.

**Comment by philip-stoev** [comment]
Here is a different message from the kernel OOM killer. I think it depends on whether the docker container is running with an explicit memory constraint or not. We explicitly constrain the memory of containers in a few tests, most notably the test/bounded-memory job in Nightly.

```
Feb 14 00:42:26 ip-10-61-32-248.ec2.internal kernel: Memory cgroup out of memory: Killed process 1415 (clusterd) total-vm:8616108kB, anon-rss:3640624kB, file-rss:47572kB, shmem-rss:0kB, UID:999 pgtables:13420kB oom_score_adj:0
```

**Comment by philip-stoev** [comment]
Here is a more detailed proposal:

- Have this entire thing in python and python only, no bash
- Perform the processing in two steps: 

Step A. Process all the logs using a list of general regexps that fish for **more-general** terms, such as "panicked" or "Out of memory" and extract all the matching log lines. Store and display all of them in the summary, regardless of whether they are known issues or not.

Step B. Only after we have collected all the log lines, if there are any log lines, try to figure out which ones are known and which ones are not, as follows:
1. Initiate a search on github via their API to fetch all tickets that have the tag `ci-flake`. (https://docs.github.com/en/rest/search?apiVersion=2022-11-28#search-issues-and-pull-requests)
2. The search results already contain all the matching tickets *and* their bodies.
3. We will annotate the important `ci-flake` bugs with a specially-formated marker in the body of the ticket. e.g. `buildkite-regexp:` . Everying between the marker and the next newline is the **more-specific** string/regexp to match
4. We apply the regexps from the tickets against the log lines and possibly obtain a matching github issue ID for each log line. Report that github issue id and also report its status (also available from the search API results from Step 1) . In the report, emphasize any issues where the status of the issue is different from resolved.

The Github API has a limit of 30 searches per minute. Given that most jobs will not result in any suspicious log lines that need to be checked, only jobs that have a failure will hit the API. Therefore, at this time there is no need to cache the REST response on our side -- each failing job can fetch the `ci-flake` issues independently and make its own determination.
